osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs

```diff
@@ -47,6 +49,34 @@ public class DrainingHealthProcessor : HealthProcessor
         private double targetMinimumHealth;
         private double drainRate = 1;
 
+        private readonly List<(double startTime, double endTime)> nonDrainSections = new List<(double, double)>();
+        private int currentNonDrainSection;
+
+        private bool isInNonDrainSection
+        {
+            get
+            {
+                if (nonDrainSections.Count == 0)
+                    return false;
+
+                var time = Time.Current;
+
+                if (time > nonDrainSections[currentNonDrainSection].endTime)
+                {
+                    while (time > nonDrainSections[currentNonDrainSection].endTime && currentNonDrainSection < nonDrainSections.Count - 1)
+                        currentNonDrainSection++;
+                }
+                else
+                {
+                    while (time < nonDrainSections[currentNonDrainSection].startTime && currentNonDrainSection > 0)
+                        currentNonDrainSection--;
+                }
+
+                var closestSection = nonDrainSections[currentNonDrainSection];
+                return time >= closestSection.startTime && time <= closestSection.endTime;
+            }
+        }
```

Comment:
* Having this much logic in a getter is a large stretch. Should be a method instead IMO.
* I'm thinking this might be a decent use case for a linked list. Then you can get rid of the index and just keep a reference to the current node.
* This is pretty much a copy of the core part of the `BreakTracker` implementation. I'd say this particular logic (determining if current time is within an interval, optimised for sequential access) needs to be shared. (Strongly encourage composition over inheritance if this happens.)

Reply:
Does making a `RangeList` implementation and a method inside to do this check sound good?

Reply:
Not sure about the name (maybe `IntervalList`?) but yeah, sounds like a reasonable way forward.

Reply:
though keep in mind it's not gonna mention times in it since this can be applied to simply any value ranges.

Reply:
*interval* is not a strictly time-specific term.

Reply:
@bdach `LinkedList` turns out to be limiting a bit of good behaviours on edge cases, for example: if the current node has been removed from the list, `Next` and `Previous` will also become `null` which means I would need to set the current node back to the first element in the list, while using indices in the other hand, the index will already point to the next element to the removed one, and I can clamp the index between `0` and the length of the list to ensure that it isn't already out-of-range.

Reply:
I'm not sure why removal of nodes is relevant here as I envisioned this being a read-only structure, but fine I suppose.